library/core/src/array/mod.rs

```diff
@@ -22,6 +22,101 @@ mod iter;
 #[stable(feature = "array_value_iter", since = "1.51.0")]
 pub use iter::IntoIter;
 
+/// Creates an array `[T; N]` where each array element `T` is returned by the `cb` call.
+///
+/// # Arguments
+///
+/// * `cb`: Callback where the passed argument is the current array index.
+///
+/// # Example
+///
+/// ```rust
+/// #![feature(array_from_fn)]
+///
+/// let array = core::array::from_fn(|i| i);
+/// assert_eq!(array, [0, 1, 2, 3, 4]);
+/// ```
+#[inline]
+#[unstable(feature = "array_from_fn", issue = "none")]
+pub fn from_fn<C, T, const N: usize>(cb: C) -> [T; N]
+where
+    C: FnMut(usize) -> T,
+{
+    // SAFETY: we know for certain that this iterator will yield exactly `N`
+    // items.
+    unsafe { collect_into_array_unchecked(&mut (0..N).map(cb)) }
+}
+
+/// Creates an array `[T; N]` where each fallible array element `T` is returned by the `cb` call.
+/// Unlike `core::array::from_fn`, where the element creation can't fail, this version will return an error
+/// if any element creation was unsuccessful.
+///
+/// # Arguments
+///
+/// * `cb`: Callback where the passed argument is the current array index.
+///
+/// # Example
+///
+/// ```rust
+/// #![feature(array_from_fn)]
+///
+/// #[derive(Debug, PartialEq)]
+/// enum SomeError {
+///     Foo,
+/// }
+///
+/// let array = core::array::try_from_fn(|i| Ok::<_, SomeError>(i));
+/// assert_eq!(array, Ok([0, 1, 2, 3, 4]));
+///
+/// let another_array = core::array::try_from_fn::<_, SomeError, (), 2>(|_| Err(SomeError::Foo));
+/// assert_eq!(another_array, Err(SomeError::Foo));
+/// ```
+#[inline]
+#[unstable(feature = "array_from_fn", issue = "none")]
+pub fn try_from_fn<C, E, T, const N: usize>(mut cb: C) -> Result<[T; N], E>
+where
+    C: FnMut(usize) -> Result<T, E>,
+{
+    struct Guard<T, const N: usize> {
+        dst: *mut T,
+        initialized: usize,
+    }
+
+    impl<T, const N: usize> Drop for Guard<T, N> {
+        fn drop(&mut self) {
+            debug_assert!(self.initialized <= N);
+
+            let initialized_part = crate::ptr::slice_from_raw_parts_mut(self.dst, self.initialized);
+            // SAFETY: this raw slice will contain only initialized objects
+            // that's why, it is allowed to drop it.
+            unsafe {
+                crate::ptr::drop_in_place(initialized_part);
+            }
+        }
+    }
+
+    let mut array: [MaybeUninit<T>; N] = MaybeUninit::uninit_array();
+    let mut guard: Guard<T, N> =
+        Guard { dst: MaybeUninit::slice_as_mut_ptr(&mut array), initialized: 0 };
+    for (idx, value) in array.iter_mut().enumerate() {
+        match cb(idx) {
+            Err(e) => {
+                // FIXME: Convert to crate::mem::transmute once it works with generics.
+                mem::forget(guard);
+                return Err(e);
+            }
+            Ok(rslt) => {
+                value.write(rslt);
+                guard.initialized += 1;
+            }
```

Comment:
I don't think that `mem::forget` here is correct - I would think the error case we will want to drop the guard and free the partially-initialized section of the array. Otherwise we might leak allocations owned by the underlying `T`s.

The good news is that without this `mem::forget` we can use a much more concise form:

```suggestion
        let item = cb(idx)?;
        value.write(item);
        guard.initialized += 1;
```

Reply:
Holy cow! When the closure returns `Err`, the partially initialized array won't be dropped 🙁
Thank you very much for pointing out this bug!

Reply:
Hey @davidhewitt. Thanks again for finishing that PR about arrays on the PyO3 repository.

Since the code is now using shared battle-tested functions that are in production today, I will resolve your concerns.